src/platform/datapath_iouring.c

```diff
@@ -0,0 +1,2128 @@
+            PktInfo->ipi_addr = SendData->LocalAddress.Ipv4.sin_addr;
+        } else {
+            Mhdr->msg_controllen += CMSG_SPACE(sizeof(struct in6_pktinfo));
+            CMsg = CXPLAT_CMSG_NXTHDR(CMsg);
+            CMsg->cmsg_level = IPPROTO_IPV6;
+            CMsg->cmsg_type = IPV6_PKTINFO;
+            CMsg->cmsg_len = CMSG_LEN(sizeof(struct in6_pktinfo));
+            struct in6_pktinfo *PktInfo6 = (struct in6_pktinfo*)CMSG_DATA(CMsg);
+            PktInfo6->ipi6_ifindex = SendData->LocalAddress.Ipv6.sin6_scope_id;
+            PktInfo6->ipi6_addr = SendData->LocalAddress.Ipv6.sin6_addr;
+        }
+    }
+
+#ifdef UDP_SEGMENT
+    if (SendData->SegmentationSupported && SendData->SegmentSize > 0 && Mhdr->msg_iov->iov_len > SendData->SegmentSize) {
+        Mhdr->msg_controllen += CMSG_SPACE(sizeof(uint16_t));
+        CMsg = CXPLAT_CMSG_NXTHDR(CMsg);
+        CMsg->cmsg_level = SOL_UDP;
+        CMsg->cmsg_type = UDP_SEGMENT;
+        CMsg->cmsg_len = CMSG_LEN(sizeof(uint16_t));
+        *((uint16_t*)CMSG_DATA(CMsg)) = SendData->SegmentSize;
+    }
+#endif
+
+    CXPLAT_DBG_ASSERT(Mhdr->msg_controllen <= sizeof(SendData->ControlBuffer));
+    SendData->ControlBufferLength = (uint8_t)Mhdr->msg_controllen;
+}
+
+QUIC_STATUS
+CxPlatSendDataSendSegmented(
+    _In_ CXPLAT_SEND_DATA* SendData,
+    _In_ BOOLEAN AlreadyLocked
+    )
+{
+    CXPLAT_DATAPATH_PARTITION* DatapathPartition = SendData->SocketContext->DatapathPartition;
+    CXPLAT_SOCKET_CONTEXT* SocketContext = SendData->SocketContext;
+    struct io_uring_sqe* Sqe;
+    QUIC_STATUS Status = QUIC_STATUS_SUCCESS;
+
+    if (!AlreadyLocked) { // Review: can we infer this from thread ID?
+        CxPlatLockAcquire(&DatapathPartition->EventQ->Lock);
+    }
+
+    if (!CxPlatListIsEmpty(&SocketContext->TxQueue)) {
+        CxPlatListInsertTail(&SocketContext->TxQueue, &SendData->TxEntry);
+        Status = QUIC_STATUS_PENDING; // TODO clean this up.
+        goto Exit;
+    }
+
+    Sqe = CxPlatAllocSqe(DatapathPartition->EventQ);
+    if (Sqe == NULL) {
+        Status = QUIC_STATUS_PENDING; // TODO clean this up.
+        CxPlatListInsertTail(&SendData->SocketContext->TxQueue, &SendData->TxEntry);
+        goto Exit;
+    }
+
+    SendData->MsgHdr.msg_name = (void*)&SendData->RemoteAddress;
+    SendData->MsgHdr.msg_namelen = sizeof(SendData->RemoteAddress);
+    SendData->MsgHdr.msg_iov = SendData->Iovs;
+    SendData->MsgHdr.msg_iovlen = 1;
+    SendData->MsgHdr.msg_flags = 0;
+    SendData->MsgHdr.msg_control = SendData->ControlBuffer;
+    SendData->MsgHdr.msg_controllen = SendData->ControlBufferLength;
+    if (SendData->ControlBufferLength == 0) {
+        CxPlatSendDataPopulateAncillaryData(SendData, &SendData->MsgHdr);
+    } else {
+        SendData->MsgHdr.msg_controllen = SendData->ControlBufferLength;
+    }
+
+    io_uring_prep_sendmsg(Sqe, SendData->SocketContext->SocketFd, &SendData->MsgHdr, 0);
+    io_uring_sqe_set_data(Sqe, (void *)&SendData->Sqe);
```

Comment:
Shouldn't there be a io_uring_submit call around here?
I am not finding where the initial submission occurs for the send side.

Reply:
Yes and no, and in the yes case, it is indeed missing :)

In the no case, the submit happens in CxPlatSocketContextIoEventComplete.